Python compatibility guidelines (#169)
GoogleCloudPlatform/cloud-opensource-python: best_practices/README.rst

has been a long standing issue. Diamond dependency (aka dependency hell) is a
common problem that a package’s dependencies depend on different and
incompatible versions of the shared packages.

.. _Python Package Index (PyPI): https://pypi.org/

Incompatibilities between packages can occur when:

- A package makes breaking API changes and doesn't follow `Semantic Versioning`_
- A package is pinning its dependency version which is conflict with other dependencies.
- A package is depending on outdated dependencies.
- A package is depending on deprecated dependencies.

.. _Semantic Versioning: https://semver.org/

This guide is a list of best practices that Python package authors can follow
to help reduce future incompatibilities. Google-sponsored projects are expected
to follow these guidelines but other projects may benefit from them as well.

-----
Tools
-----

    To detect and prevent version incompatibilities between Google OSS Python
    packages, we provide a set of tooling to ensure we are compatible with
    ourselves. And the tooling can also run checks for any package that wants to
    know whether it is self-compatible or compatible with a list of other packages.

>     Below lists our tooling:

Author: This is for displaying this paragraph with a background.

Dependency Compatibility Dashboard
----------------------------------

The `dashboard`_ shows the compatibility status of a list of Google owned OSS
Python packages.

.. _dashboard: https://googlecloudplatform.github.io/cloud-opensource-python/

Badge Server
------------

The badge server runs checks and generate svg format badge image to show the
status of a given package. Supported usage includes:

- Self compatibility
- Google-wise compatibility
- Dependency version status
- One badge for all kinds of checks above

For more details please point to `here`_.

.. _here: https://github.com/GoogleCloudPlatform/cloud-opensource-python/tree/master/badge_server

----------------------
Python Packaging Rules
----------------------

Package Versioning
------------------

We should use semantic versioning for all Google OSS Python distribution
packages. Semantic versioning requires that given a version number
`MAJOR.MINOR.PATCH`, increment the:

    * MAJOR version when you make incompatible API changes.

>     * MINOR version when you add functionality in a backwards-compatible manner.

Author: Yes, we don't need to include it here. This is copied from the semantic version website (https://semver.org/) and we should keep it as it is. The details about breaking change detector could be in its own README file.

>     * PATCH version when you make backwards-compatible bug fixes.

Author: Same here.

Requirements:

> - Non-GA libraries should use major version 0, and be promoted to 1.0 when reach GA.

Author: Done.

> - Non-GA libraries should use major version 0, and be promoted to 1.0 when reach GA.

Author: Done.

- Non-GA libraries could be excluded from semver stability guarantees.
- GA libraries must conform to semantic versioning.

Dependencies
------------

**1. Specify dependency version using closed ranges**

> - Minor or patch version shouldn’t be used as an upper bound for 1st party dependencies unless the dependency is not GA.

Author: Done.

- Specific versions could be excluded if they are known to be incompatible. e.g. google-cloud-pubsub >= 0.1.1 !=2.0.0 !=2.0.1
- Specific versions may be specified if a package exists as a wrapper around another.

> - Version range upper bound should be updated when there is a newer version available as soon as possible.

Author: Done.

- We allow 30 days grace period for package owners to migrate to support new major version bump of the dependencies. e.g. google-cloud-dataflow may specify an exact version of apache-beam.

**2. Avoid depending on unstable release version dependencies**

> - It’s not recommended to depend on non-GA packages.

Author: Done.

> - It’s not recommended to depend on non-GA packages.

Author: Think about it more, `Avoid` is too strict for this requirement, some libraries have to depend on non-GA libraries, I'm going to change back to previous version.

> - Avoid depending on pre-release, post-release and development release version.

Author: Done.

> - Dependency version should be final release. (`PEP 440`_)

Author: Done.

> - Dependency version should be final release. (`PEP 440`_)

Author: This is actually saying the same thing as the point above, any version which is not a pre-release, post-release and development release is a final release. Removed this point.

> - GA packages must not depend on non-GA packages.

Author: Yes, mentioned this at the top of this file.

> **3. Version range upper bound should be updated when there is a newer version available as soon as possible.**

Author: Oh I was thinking making this a separate point to emphasize on it.. and forgot to delete it from the sub-point in. Removed.

> - We allow 30 days grace period for package owners to migrate to support new major version bump of the dependencies.

Author: Done.

**4. Minimize dependencies**

- Packages should use Python built-in modules if possible. e.g. logging, unittest

> **5. Never vendor dependencies**

Author: Done.

.. _PEP 440: https://www.python.org/dev/peps/pep-0440/#final-releases

Release and Support
-------------------

> - Major version bump should be rare

Author: Done.

- Minimize the cost for users to go from one major version to another.
- Support every semver-major HEAD of every package that is 1.0+ for at least one year.
- Dropping support for any older version should be semver-major.

---------------
GA Requirements
---------------

> The GA requirements are validated using the `github badge`_ service, the badge

Author: Done.

should be green before any GA launch.

> - Package must be self compatible

Author: Done.

> If package A depends on dependency B and C, and they require different version

Author: Done.

> of dependency D, package A is not self compatible. Packages that are not self

Reviewer: maybe "internally" rather than self?

Author: Done.

> compatible will have conflicts with all the rest packages in the world.

Author: Done.

> - Package must be google-wise compatible

Author: Done.

It’s required for any new package owned by Google to be compatible with all the other Google Python packages. So that using any combination of Google Python packages will not cause any conflicts during installation or failures during runtime.

> - Package must support latest version of its dependencies

Author: Done.

Python Package Compatibility Guidelines
=======================================

This document uses terminology (MUST, SHOULD, etc) from `RFC 2119`_.

.. _RFC 2119: https://www.ietf.org/rfc/rfc2119.txt

----------
Background
----------

Incompatibilities between packages published on the `Python Package Index (PyPI)`_
have been a long standing issue. Diamond dependencies are a common problem where
a package or application has dependencies that are, themselves, dependant on
incompatible versions of shared packages.

.. _Python Package Index (PyPI): https://pypi.org/

Incompatibilities between packages can occur when:

- A package makes breaking API changes and doesn't follow `Semantic Versioning`_
- A package has a pinned dependency version which conflicts with other dependencies.
- A package depends on outdated dependencies.

> - A package is dependent on deprecated dependencies.

Author: Currently the deprecated dependency finder tool in our compatibility_lib uses the development status field in setup.py to determine whether a package is deprecated or not. If one pkg is deprecated, the status should be `Development Status :: 7 - Inactive`.

.. _Semantic Versioning: https://semver.org/

This guide is a list of best practices that Python package authors can follow
to help reduce future incompatibilities. Google-sponsored projects are expected
to follow these guidelines but other projects may benefit from them as well.

-----
Tools
-----

    To detect and prevent version incompatibilities between Google Open Source Python
    packages, we provide a set of tooling to ensure we are compatible with

>     ourselves. Below lists our tooling:

Reviewer: ```suggestion
    ourselves. Our tools are as follows:
```

Dependency Compatibility Dashboard
----------------------------------

The `dashboard`_ shows the compatibility status of a list of Google owned OSS
Python packages.

.. _dashboard: https://googlecloudplatform.github.io/cloud-opensource-python/

Badge Server
------------

The badge server runs checks and generates an svg format badge image to show the
status of a given package. Supported usage includes:

- Self compatibility
- Google-wise compatibility
- Dependency version status
- One badge for all kinds of checks above

For more details please refer to `here`_.

.. _here: https://github.com/GoogleCloudPlatform/cloud-opensource-python/tree/master/badge_server

----------------------
Python Packaging Rules
----------------------

Package Versioning
------------------

We should use semantic versioning for all Google OSS Python distribution
packages. Semantic versioning requires that given a version number
`MAJOR.MINOR.PATCH`, increment the:

    * MAJOR version when you make incompatible API changes.
    * MINOR version when you add functionality in a backwards-compatible manner.
    * PATCH version when you make backwards-compatible bug fixes.

Requirements:

> - GA(Generally Available) libraries must conform to semantic versioning.

Reviewer: Make "Generally Available" a link to https://cloud.google.com/terms/launch-stages

Python Package Compatibility Guidelines
=======================================

This document uses terminology (MUST, SHOULD, etc) from `RFC 2119`_.

.. _RFC 2119: https://www.ietf.org/rfc/rfc2119.txt

----------
Background
----------

Incompatibilities between packages published on the `Python Package Index (PyPI)`_

> has been a long standing issue. Diamond dependencies are a common problem where

Reviewer: ```suggestion
have been a long standing issue. Diamond dependencies are a common problem where
```

a package or application has dependencies that are, themselves, dependant on
incompatible versions of shared packages.

> has been a long standing issue. Diamond dependency (aka dependency hell) is a

Reviewer: Did you forget to remove this?

Author: Removed.

common problem that a package’s dependencies depend on different and
incompatible versions of the shared packages.

.. _Python Package Index (PyPI): https://pypi.org/

Incompatibilities between packages can occur when:

- A package makes breaking API changes and doesn't follow `Semantic Versioning`_
- A package has a pinned dependency version which conflicts with other dependencies.
- A package depends on outdated dependencies.

> - A package is depends on deprecated dependencies.

Reviewer: ```suggestion
- A package is dependent on deprecated dependencies.
```

Is this different than the above point?

Author: Yes depending on deprecated package could still be the latest version but it's deprecated.

.. _Semantic Versioning: https://semver.org/

This guide is a list of best practices that Python package authors can follow
to help reduce future incompatibilities. Google-sponsored projects are expected
to follow these guidelines but other projects may benefit from them as well.

-----
Tools
-----

>     To detect and prevent version incompatibilities between Google OSS Python

Reviewer: ```suggestion
    To detect and prevent version incompatibilities between Google Open Source Python
```

    packages, we provide a set of tooling to ensure we are compatible with

>     ourselves. And the tooling can also run checks for any Google owned package that wants to

Reviewer: Does the sentence starting with "And..." still make sense now that only Google-supported packages are supported?

Author: Removed.

    know whether it is self-compatible or compatible with a list of other packages.
    Below lists our tooling:

Dependency Compatibility Dashboard
----------------------------------

> The `dashboard`_ shows the compatibility status of a list of Google owned OSS

Reviewer: ```suggestion
The `dashboard`_ shows the compatibility status of a list of Google sponsored Open Source
```

Python packages.

.. _dashboard: https://googlecloudplatform.github.io/cloud-opensource-python/

Badge Server
------------

The badge server runs checks and generates an svg format badge image to show the
status of a given package. Supported usage includes:

> - Self compatibility

Reviewer: ```suggestion
- Self compatibility (the package has internally consistent dependencies)
```

> - Google-wise compatibility

Reviewer: ```suggestion
- Google-wise compatibility (the package is compatible with other Google packages)
```

> - Dependency version status

Reviewer: ```suggestion
- Dependency version status (the package does not rely on obsolete dependencies)
```

> - One badge for all kinds of checks above

Reviewer: ```suggestion
- One badge for all of the checks above
```
